File: packages/sdk/src/integrations/openai.ts
/**
 * AgentGuard — OpenAI Integration
 *
 * Wraps an OpenAI client instance to intercept function/tool calls returned
 * in chat completion responses, evaluating each through AgentGuard before
 * the caller can execute them.
 *
 * Usage:
 *   import { openaiGuard } from '@the-bot-club/agentguard';
 *   import OpenAI from 'openai';
 *
 *   const client = new OpenAI({ apiKey: '...' });
 *   const guarded = openaiGuard(client, { apiKey: 'ag_...' });
 *
 *   // Use `guarded` exactly like the OpenAI client:
 *   const response = await guarded.chat.completions.create({ ... });
 *
 *   // response._agentguard contains per-call decisions:
 *   if (response._agentguard?.hasBlocks) {
 *     // One or more tool calls were blocked — do not execute them
 *   }
 *
 * The guard attaches a `_agentguard` field to each response that contains
 * tool calls. Check this field before executing tool calls.
 */
import { AgentGuard } from '../sdk/client.js';
import { AgentGuardBlockError } from './errors.js';

// ─── Options ──────────────────────────────────────────────────────────────────

export interface OpenAIGuardOptions {
  /** AgentGuard API key (ag_...) */
  apiKey: string;
  /** Override the AgentGuard API base URL */
  baseUrl?: string;
  /** Optional agent ID for scoped evaluations */
  agentId?: string;
  /**
   * When true, throw AgentGuardBlockError immediately if any tool call is blocked.
   * When false (default), attach results to response._agentguard and let caller handle.
   */
  throwOnBlock?: boolean;
}

// ─── Minimal OpenAI type stubs (structural, no hard dep on openai package) ────

interface ToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

interface ChatCompletionMessage {
  role: string;
  content: string | null;
  tool_calls?: ToolCall[];
}

interface ChatCompletionChoice {
  index: number;
  message: ChatCompletionMessage;
  finish_reason: string;
}

interface ChatCompletionResponse {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: ChatCompletionChoice[];
  /** Attached by AgentGuard — not part of the OpenAI spec */
  _agentguard?: AgentGuardBatchResult;
}

// ─── AgentGuard batch result shape ────────────────────────────────────────────

export interface AgentGuardToolDecision {
  /** 0-based index matching the tool_calls array */
  index: number;
  /** Tool name */
  tool: string;
  /** Policy decision */
  decision: 'allow' | 'block' | 'monitor' | 'require_approval';
  /** Risk score 0–1000 */
  riskScore: number;
  /** Human-readable block reason (present on block/require_approval) */
  reason?: string;
  /** What to do instead (present on block/require_approval) */
  suggestion?: string;
  /** Docs link (present on block/require_approval) */
  docs?: string;
  /** Alternative tools (present on block) */
  alternatives?: string[];
  /** Matched policy rule ID */
  matchedRuleId?: string;
  /** Approval URL (present when decision is require_approval) */
  approvalUrl?: string;
}

export interface AgentGuardBatchResult {
  /** Individual decisions, parallel to tool_calls array */
  decisions: AgentGuardToolDecision[];
  /** True if any decision is "block" */
  hasBlocks: boolean;
  /** True if any decision is "require_approval" */
  hasApprovals: boolean;
  /** Blocked tool call indices */
  blockedIndices: number[];
  /** Summary counts */
  summary: {
    total: number;
    allowed: number;
    blocked: number;
    monitored: number;
    requireApproval: number;
  };
}

// ─── Minimal OpenAI client structural interface ────────────────────────────────

interface OpenAIClientLike {
  chat: {
    completions: {
      create(params: unknown): Promise<ChatCompletionResponse>;
    };
  };
  [key: string]: unknown;
}

// ─── openaiGuard ──────────────────────────────────────────────────────────────

/**
 * Wrap an OpenAI client with AgentGuard policy enforcement.
 *
 * Intercepts `chat.completions.create` responses that contain tool_calls,
 * evaluating each tool call through AgentGuard before returning the response.
 *
 * One-liner:
 * ```typescript
 * import { openaiGuard } from '@the-bot-club/agentguard';
 * const guarded = openaiGuard(new OpenAI({ apiKey: '...' }), { apiKey: 'ag_...' });
 * ```
 *
 * @param client   An OpenAI client instance (or any structurally compatible object)
 * @param options  AgentGuard configuration
 * @returns        A wrapped client with the same interface
 */
export function openaiGuard<T extends OpenAIClientLike>(
  client: T,
  options: OpenAIGuardOptions,
): T & { chat: { completions: { create(params: unknown): Promise<ChatCompletionResponse> } } } {
  const guard = new AgentGuard({
    apiKey: options.apiKey,
    baseUrl: options.baseUrl,
  });

  const throwOnBlock = options.throwOnBlock ?? false;

  // Build the intercepted create function
  const guardedCreate = async (params: unknown): Promise<ChatCompletionResponse> => {
    // Call the real OpenAI API
    const response = await client.chat.completions.create(params);

    // Only intercept if the response contains tool_calls
    const toolCalls = response.choices?.[0]?.message?.tool_calls;
    if (!toolCalls || toolCalls.length === 0) {
      return response;
    }

    // Evaluate each tool call through AgentGuard
    const decisions: AgentGuardToolDecision[] = await Promise.all(
      toolCalls.map(async (tc, idx) => {
        let parsedArgs: Record<string, unknown> = {};
        try {
          parsedArgs = JSON.parse(tc.function.arguments || '{}') as Record<string, unknown>;
        } catch {
          parsedArgs = { _raw: tc.function.arguments };
        }

        const result = await guard.evaluate({
          tool: tc.function.name,
          params: parsedArgs,
        });

        // Normalise result shape (API returns `result` field, not `decision`)
        const raw = result as unknown as Record<string, unknown>;
        const decision = (raw['result'] ?? raw['decision'] ?? 'allow') as AgentGuardToolDecision['decision'];

        return {
          index: idx,
          tool: tc.function.name,
          decision,
          riskScore: (raw['riskScore'] as number | undefined) ?? 0,
          reason: raw['reason'] as string | undefined,
          suggestion: raw['suggestion'] as string | undefined,
          docs: raw['docs'] as string | undefined,
          alternatives: raw['alternatives'] as string[] | undefined,
          matchedRuleId: raw['matchedRuleId'] as string | undefined,
          approvalUrl: raw['approvalUrl'] as string | undefined,
        };
      }),
    );

    // Build summary
    const blocked = decisions.filter((d) => d.decision === 'block');
    const summary: AgentGuardBatchResult['summary'] = {
      total: decisions.length,
      allowed: decisions.filter((d) => d.decision === 'allow').length,
      blocked: blocked.length,
      monitored: decisions.filter((d) => d.decision === 'monitor').length,
      requireApproval: decisions.filter((d) => d.decision === 'require_approval').length,
    };

    const batchResult: AgentGuardBatchResult = {
      decisions,
      hasBlocks: blocked.length > 0,
      hasApprovals: summary.requireApproval > 0,
      blockedIndices: decisions.filter((d) => d.decision === 'block').map((d) => d.index),
      summary,
    };

    // Attach guard results to the response
    response._agentguard = batchResult;

    // Throw on first block if requested
    if (throwOnBlock && blocked.length > 0) {
      const firstBlock = blocked[0]!;
      throw new AgentGuardBlockError({
        ...firstBlock,
        agentId: options.agentId,
      });
    }

    return response;
  };

  // Return a proxy that preserves all original client properties
  return {
    ...client,
    chat: {
      ...((client.chat as Record<string, unknown>) ?? {}),
      completions: {
        ...((client.chat?.completions as Record<string, unknown>) ?? {}),
        create: guardedCreate,
      },
    },
  } as T & { chat: { completions: { create(params: unknown): Promise<ChatCompletionResponse> } } };
}

File: packages/python/agentguard/integrations/openai.py
"""AgentGuard — OpenAI Integration (Python).

Wraps an OpenAI client instance to intercept function/tool calls returned
in chat completion responses, evaluating each through AgentGuard before
the caller can execute them.

Usage::

    from openai import OpenAI
    from agentguard.integrations import openai_guard

    client = OpenAI(api_key="sk-...")
    guarded = openai_guard(client, api_key="ag_...")

    response = guarded.chat.completions.create(
        model="gpt-4o",
        messages=[...],
        tools=[...],
    )

    # response has _agentguard attached with per-call decisions
    if response._agentguard and response._agentguard["has_blocks"]:
        # One or more tool calls were blocked — do not execute them
        for decision in response._agentguard["decisions"]:
            if decision["decision"] == "block":
                print(f"Blocked {decision['tool']}: {decision['reason']}")

The guard attaches an ``_agentguard`` attribute to responses that contain
tool calls. Check this before executing the tool calls.
"""
import json
from typing import Any, Dict, List, Optional

from ..client import AgentGuard
from .errors import AgentGuardBlockError


def openai_guard(
    client: Any,
    api_key: str,
    base_url: str = "https://api.agentguard.tech",
    agent_id: Optional[str] = None,
    throw_on_block: bool = False,
) -> Any:
    """Wrap an OpenAI client with AgentGuard policy enforcement.

    Intercepts ``chat.completions.create`` responses that contain tool_calls,
    evaluating each tool call through AgentGuard before returning the response.

    One-liner::

        from agentguard.integrations import openai_guard
        guarded = openai_guard(client, api_key="ag_...")

    Args:
        client:         An OpenAI client instance (``openai.OpenAI`` or compatible)
        api_key:        AgentGuard API key (``ag_...``)
        base_url:       Optional override for the AgentGuard API base URL
        agent_id:       Optional agent ID for scoped evaluations
        throw_on_block: If True, raise ``AgentGuardBlockError`` on first block
                        instead of attaching results to the response object.

    Returns:
        A wrapped client with the same interface as the original, with
        ``chat.completions.create`` intercepted.
    """
    guard = AgentGuard(api_key=api_key, base_url=base_url)

    class _GuardedCompletions:
        """Proxy that wraps the original completions interface."""

        def __init__(self, original: Any) -> None:
            self._original = original

        def create(self, **kwargs: Any) -> Any:
            response = self._original.create(**kwargs)

            # Only intercept responses that have tool_calls
            tool_calls = _extract_tool_calls(response)
            if not tool_calls:
                return response

            # Evaluate each tool call through AgentGuard
            decisions: List[Dict[str, Any]] = []
            for idx, tc in enumerate(tool_calls):
                tool_name = _get_tool_name(tc)
                raw_args = _get_tool_args(tc)
                try:
                    parsed_args: Dict[str, Any] = json.loads(raw_args) if isinstance(raw_args, str) else (raw_args or {})
                except (json.JSONDecodeError, ValueError):
                    parsed_args = {"_raw": raw_args}

                result = guard.evaluate(tool=tool_name, params=parsed_args)
                decision_val = result.get("result", result.get("decision", "allow"))

                decision: Dict[str, Any] = {
                    "index": idx,
                    "tool": tool_name,
                    "decision": decision_val,
                    "risk_score": result.get("riskScore", result.get("risk_score", 0)),
                    "reason": result.get("reason"),
                    "suggestion": result.get("suggestion"),
                    "docs": result.get("docs"),
                    "alternatives": result.get("alternatives", []),
                    "matched_rule_id": result.get("matchedRuleId", result.get("matched_rule_id")),
                    "approval_url": result.get("approvalUrl", result.get("approval_url")),
                }
                decisions.append(decision)

            # Build summary
            blocked = [d for d in decisions if d["decision"] == "block"]
            summary = {
                "total": len(decisions),
                "allowed": sum(1 for d in decisions if d["decision"] == "allow"),
                "blocked": len(blocked),
                "monitored": sum(1 for d in decisions if d["decision"] == "monitor"),
                "require_approval": sum(1 for d in decisions if d["decision"] == "require_approval"),
            }

            agentguard_result = {
                "decisions": decisions,
                "has_blocks": len(blocked) > 0,
                "has_approvals": summary["require_approval"] > 0,
                "blocked_indices": [d["index"] for d in blocked],
                "summary": summary,
            }

            # Attach to response (works for both OpenAI and mock objects)
            try:
                object.__setattr__(response, "_agentguard", agentguard_result)
            except (AttributeError, TypeError):
                try:
                    response._agentguard = agentguard_result  # type: ignore[attr-defined]
                except (AttributeError, TypeError):
                    pass

            # Throw on first block if requested
            if throw_on_block and blocked:
                raise AgentGuardBlockError({**blocked[0], "agent_id": agent_id})

            return response

    class _GuardedChat:
        """Proxy that wraps the original chat interface."""

        def __init__(self, original: Any) -> None:
            self._original = original
            self.completions = _GuardedCompletions(original.completions)

        def __getattr__(self, name: str) -> Any:
            return getattr(self._original, name)

    class _GuardedClient:
        """Proxy wrapper around the original OpenAI client."""

        def __init__(self, original: Any) -> None:
            self._original = original
            self.chat = _GuardedChat(original.chat)

        def __getattr__(self, name: str) -> Any:
            return getattr(self._original, name)

    return _GuardedClient(client)


# ─── Helpers ──────────────────────────────────────────────────────────────────

def _extract_tool_calls(response: Any) -> List[Any]:
    """Extract tool_calls from an OpenAI response object."""
    try:
        choices = response.choices
        if choices and len(choices) > 0:
            msg = choices[0].message
            return list(msg.tool_calls or [])
    except AttributeError:
        pass
    # Dict-style response fallback
    if isinstance(response, dict):
        choices = response.get("choices", [])
        if choices:
            msg = choices[0].get("message", {})
            return msg.get("tool_calls", []) or []
    return []


def _get_tool_name(tool_call: Any) -> str:
    """Extract tool name from a tool_call object or dict."""
    try:
        return tool_call.function.name
    except AttributeError:
        pass
    if isinstance(tool_call, dict):
        return tool_call.get("function", {}).get("name", "unknown")
    return "unknown"


def _get_tool_args(tool_call: Any) -> Any:
    """Extract tool arguments from a tool_call object or dict."""
    try:
        return tool_call.function.arguments
    except AttributeError:
        pass
    if isinstance(tool_call, dict):
        return tool_call.get("function", {}).get("arguments", "{}")
    return "{}"
